[meglwhy] iP

meglwhy opened the pull request:
### **Koji**

Koji is a smart, efficient, and easy-to-use task manager that helps you stay organized. It's:
- Text-based
- Easy to learn
- FAST & SUPER EFFICIENT
---

**Getting Started**
1. Download Koji from [this link](https://github.com/meglwhy/ip).
2. Double-click the .jar file.
3. Start adding tasks to organize your life.
4. Enjoy the convenience! 😉
 
 **Features**
- Managing tasks (ToDo, Deadline, Event)
- Tracking deadlines
- Finding tasks by keywords
- Deleting tasks
- Saving tasks automatically
- More features coming soon!
 
---
If you are a Java programmer, you can use it to practice Java too. Here's the main method:

```java
public class Main {
    public static void main(String[] args) {
        System.out.println("Welcome to Koji!");
        new Koji("data/tasks.txt").run();
    }
}
``` 
 ---
 **To-Do List**

- [ ]  Implement Todo, Deadline, Event tasks
- [ ]  Add find command 🔍
- [ ]  Implement reminders (coming soon) ⏳
- [ ]  Improve UI & design 🎨

 ---
 **Example Commands**
```
todo read book
deadline return book /by 2023-12-31 1800
event project meeting /from 2023-10-01 1400 /to 2023-10-01 1600
find book
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

meglwhy committed:
level 0

meglwhy committed:
level 1

meglwhy committed:
level 2

meglwhy committed:
level 3

meglwhy committed:
level 4

meglwhy committed:
a-textuitesting

meglwhy committed:
level 5

meglwhy committed:
level 6

Jeremiah33-3 reviewed (commented):
Overall, the naming of variables and methods are easy to understand, good job on that! 👍 Jiayouuu

dexterleng commented on src/main/java/Koji.java:
```diff
@@ -0,0 +1,252 @@
+        if (!input.contains("/from") || !input.contains("/to")) {
+            throw new KojiException(" what event u doin");
+        }
+        String[] parts = input.substring(6).split(" /from | /to ", 3);
```
Something like `event /from something /from somethingElse` will pass validation (the second /from will be treated as /to). Try using a more strict validation logic.

dexterleng commented on src/main/java/TaskList.java:
```diff
@@ -0,0 +1,34 @@
+    }
+
+    public Task delete(int index) throws KojiException {
+        return tasks.remove(index);
```
`ArrayList#remove()` does not throw a `KojiException`.

Consider checking the index here before doing the remove so you can throw a KojiException with a descriptive message.

dexterleng commented:
Good job! Just a few minor nits but coding style is on point.

dexterleng commented on src/main/java/Koji.java:
```diff
@@ -0,0 +1,252 @@
+import java.util.*;
+
+//public class Koji {
```
Delete the commented code. You can always restore it from your git history if you need to.

NHT020305 commented on src/main/java/Event.java:
```diff
@@ -0,0 +1,15 @@
+public class Event extends Task {
```
should we put header comments for this public class

NHT020305 commented on src/main/java/Koji.java:
```diff
@@ -0,0 +1,252 @@
+//    }
+//
+//    private static void updateTaskStatus(String input, boolean markAsDone) {
+//            int taskIndex = Integer.parseInt(input.split(" ")[1]) - 1;
```
Should we delete the commented code ?

NHT020305 reviewed (commented):
LGTM. Just a few nits to fix.

zechary28 commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,27 @@
+public class Task {
```
I don't see a need for Task to be instantiated, maybe can make this an abstract class

zechary28 commented on src/main/java/TaskList.java:
```diff
@@ -0,0 +1,34 @@
+
+    public int size() {
+        return tasks.size();
+    }
```
I would personally name the method a verb getSize() to be clearer and make a distinction between variables and methods

zechary28 reviewed (commented):
overall pretty good, i see you're a bit behind, i think week 2. but you've got a good base to work on. jiayous!! lets get the full marks for ip rather than less than half

NHT020305 commented on src/main/java/Koji.java:
```diff
@@ -0,0 +1,252 @@
+import java.util.*;
```
I think imported packages should be listed explicitly

NHT020305 reviewed (commented):
LGTM. Just a few nits to fix.

meglwhy committed:
Refactor project structure: Organize files into koji and task packages

meglwhy committed:
Add Storage class to save tasks in duke.txt

meglwhy committed:
Merge branch-Level-7

meglwhy committed:
Use LocalDateTime for parsing and formatting dates in Deadline and Event

meglwhy committed:
Merge branch-Level-8 (corrected)

meglwhy committed:
Refactor project structure into more OOP design

meglwhy committed:
Delete commented lines and fix event date parsing and storage format

meglwhy committed:
Add JavaDoc comments to the code

meglwhy committed:
Enforce coding standards across all classes

meglwhy committed:
Implement find feature

meglwhy committed:
Add Gradle to project section

meglwhy committed:
Correct implementation of Gradle

meglwhy committed:
Add JUnit tests

meglwhy committed:
Package the app as an executable JAR file

meglwhy committed:
Implement (incomplete) version of checkstyle

meglwhy committed:
Remove debugging statements

meglwhy committed:
Implement complete GUI layout with JavaFX controls and custom DialogBox

meglwhy committed:
Create a GUI for Koji

meglwhy committed:
Use the JavaFX technology to implement the GUI (completed)

meglwhy reviewed (commented):
Create a GUI for Koji

This commit sets up the initial GUI for the Koji chatbot. The entry point
has been updated to launch a JavaFX application that will serve as the main
interface for the chatbot. Basic components and structure are now in place,
preparing the project for interactive features.

meglwhy committed:
Add assertions to document key assumptions in the code

Added Java assert statements in critical methods and constructors to ensure that important assumptions hold during execution. For example, assertions now verify that:
- Input values are not null.
- Parsed dates in Deadline and Event classes are valid (non-null).
- Task objects added to the TaskList are not null.

These runtime checks help catch logical errors early when assertions are enabled (with the -ea flag), serving both as documentation and as a debugging aid.

meglwhy committed:
Merge branch 'branch-A-Assertions'

Add assertions to document key assumptions

This merge integrates changes that add Java assert statements in critical parts of
the codebase. Assertions have been added to ensure that:
- Input values are not null.
- Parsed dates in Deadline and Event classes are valid.
- Task objects added to the TaskList are not null.

These runtime checks help catch logical errors early (when assertions are enabled)
and serve as both documentation and a debugging aid.

meglwhy committed:
Remove unused/duplicate DialogBox class

meglwhy committed:
Refactor DialogBox: Extract FXML loading and improve documentation

Refactored the DialogBox class to enhance readability and maintainability.
- Extracted the FXML loading logic into a separate loadFXML() method.
- Enhanced JavaDoc comments for clarity.
- Retained functionality for creating user and system dialog boxes,
  including flipping the dialog for system messages.
This change cleans up the constructor and simplifies future modifications.

meglwhy committed:
Merge branch 'branch-A-CodeQuality'
Improve code quality and enforce coding standards

This merge incorporates changes from branch-A-CodeQuality, which refactors the code for better
readability, adds assertions to document key assumptions, and aligns code formatting with our
quality guidelines. These improvements enhance maintainability and help catch bugs earlier.

meglwhy committed:
Refactor TaskList.findTasks to use Java Streams

This commit replaces the explicit for-loop in the TaskList.findTasks method with a stream-based approach. The new implementation leverages the Java Stream API to filter tasks by a keyword in a case-insensitive manner.

meglwhy committed:
Refactor TaskList.findTasks to use Java Streams

This commit replaces the explicit for-loop in the TaskList.findTasks method with a stream-based approach. The new implementation leverages the Java Stream API to filter tasks by a keyword in a case-insensitive manner.

meglwhy committed:
Revert "Refactor TaskList.findTasks to use Java Streams"

This reverts commit 2417f13e2fad0d6866497e759db61f3a41f3d97c.

meglwhy committed:
Refactor TaskList.findTasks to use Java Streams

This commit replaces the explicit for-loop in the TaskList.findTasks method with a stream-based approach. The new implementation leverages the Java Stream API to filter tasks by a keyword in a case-insensitive manner.

meglwhy committed:
Merge branch-A-Streams into master

meglwhy committed:
Add comments for change

meglwhy committed:
Merge pull request #4 from meglwhy/master

Merge branch 'branch-A-CodeQuality'

meglwhy committed:
Merge branch 'branch-A-CodeQuality' of https://github.com/meglwhy/ip into branch-A-CodeQuality

meglwhy committed:
Delete duplicate DialogBox class

meglwhy committed:
Add assertions to document key assumptions in the code

Added Java assert statements in critical methods and constructors to ensure that important assumptions hold during execution. For example, assertions now verify that:
- Input values are not null.
- Parsed dates in Deadline and Event classes are valid (non-null).
- Task objects added to the TaskList are not null.

These runtime checks help catch logical errors early when assertions are enabled (with the -ea flag), serving both as documentation and as a debugging aid.

meglwhy committed:
Merge pull request #5 from meglwhy/branch-A-Assertions

Merge branch 'branch-A-Assertions'

meglwhy committed:
Refactor DialogBox class for improved readability and maintainability

meglwhy committed:
Revert "Refactor DialogBox class for improved readability and maintainability"

This reverts commit ed0242c367b2d9c31b774805d48cf3300fa896dd.

meglwhy committed:
Merge branch 'master' into branch-A-CodeQuality

meglwhy committed:
Merge pull request #6 from meglwhy/branch-A-CodeQuality

Merge branch 'branch-A-CodeQuality'

meglwhy committed:
Merge branch 'master' of https://github.com/meglwhy/ip

meglwhy committed:
Refactor task search to use Java Streams

Refactored the findTasks method in TaskList to utilize Java Streams instead
of a traditional for-loop. This change simplifies the code and improves
readability by making the filtering logic more declarative. The new
implementation filters tasks based on the search keyword and collects the
results into an ArrayList.

This approach not only reduces boilerplate but also lays the foundation for
potential future enhancements, such as parallel processing, if needed.

meglwhy committed:
Merge pull request #7 from meglwhy/branch-A-Streams

Refactor task search to use Java Streams

meglwhy committed:
Set up CI with GitHub Actions for Gradle build

meglwhy committed:
Add edit command feature to update tasks

meglwhy committed:
Refactor command classes to return responses

Updated all command classes so that their execute() methods now return a String response.
This allows the GUI to display responses generated by the commands, providing better
integration between user input and system output.

meglwhy committed:
Improve GUI interactivity and layout for Koji chatbot

Added event handlers for user input, auto-scrolling for the dialog container, and updated layout constraints to center the ScrollPane and prevent horizontal scrolling.
Inserted a welcome message on startup.
Updated DialogBox styling to add padding between text and avatar and to display avatars as circular images.
These improvements enhance the overall user experience.

meglwhy committed:
Improve GUI layout and implement mark/unmark commands

Enhanced the chatbot GUI by refining the layout to ensure that the dialog
container auto-scrolls and displays full-length messages without truncation.
Adjusted component sizes, spacing, and background styling for a more polished
look.

Implemented mark and unmark commands to allow users to update task statuses
in-place. Updated the commands and TaskList methods to use 1-based indexing,
ensuring that commands like 'mark 6' and 'unmark 6' update the correct tasks.
These improvements collectively enhance the user experience by providing
both a more visually appealing interface and interactive task updates.

meglwhy committed:
Refactor exit command and update build.gradle for Windows compatibility

Modified ExitCommand to call Platform.exit() so that when the user types 'bye', the application window is closed immediately. This change ensures a clean shutdown of the GUI.

Also updated build.gradle to work correctly on Windows by adjusting the JavaFX dependency declarations. Platform-specific suffixes have been set appropriately to avoid build issues on Windows environments.

These changes improve the overall user experience and enhance cross-platform compatibility.

meglwhy committed:
Add event date validation to ensure start date is before end date

Updated AddEventCommand to validate event dates before creating an Event task.
Now the command parses the start and end dates using Task.parseDate, checks that
both dates are valid, and ensures that the start date is not after the end date.
If the validation fails, the command returns a descriptive error message.
This change improves data integrity by preventing events with invalid date ranges.

meglwhy committed:
Refactor chatbot responses for consistency and improved clarity

Updated the command execution flow so that each command now returns a response
string instead of directly printing to the UI. This refactoring applies to the
ExitCommand, ListCommand, and other commands, enabling the UI to capture and display
responses uniformly. These changes improve separation of concerns and pave the way
for future enhancements in response handling.

meglwhy committed:
Update User Guide with comprehensive feature documentation and GitHub Pages setup

Updated docs/README.md to provide detailed instructions for using Koji Chatbot. The guide now covers how to add deadlines, mark/unmark tasks, edit tasks, find tasks, add events, and delete tasks. It also explains the command format and includes step-by-step instructions for enabling GitHub Pages for hosting the user guide. These enhancements help users better understand and interact with Koji.

meglwhy committed:
docs: expand README documentation

Added more details about project setup and usage instructions.

meglwhy committed:
Improve README documentation

Added more details about project setup and usage instructions.

meglwhy committed:
Change dependencies to accomodate for linux

meglwhy committed:
Add JUnit 5 dependencies for unit testing

meglwhy committed:
Fix DeadlineTest to match OUTPUT_FORMAT in Task

meglwhy committed:
Fix testDeadlineToString formatting issue

Deadline.toString() was producing a different format than expected in
testDeadlineToString. The expected format was '(by: 02 Mar 2025, 07:00 PM)',
but the actual output was incorrect due to the OUTPUT_FORMAT used in Task.

The OUTPUT_FORMAT in Task was using 'MMM dd yyyy, h:mm a', which did not
ensure leading zeros for single-digit hours and had the wrong order for
day and month.

Let's update OUTPUT_FORMAT to 'dd MMM yyyy, hh:mm a' to ensure:
- Day appears before the month.
- Hours are displayed with leading zeros in 12-hour format.
- The AM/PM notation is correctly formatted.

This fix ensures that testDeadlineToString now passes and matches the
expected output format.

meglwhy committed:
File creation: Add warning message for file existence check

When creating a new file in the Storage class, the code silently proceeds
even if the file already exists or cannot be created. This lack of
feedback can make debugging file-related issues difficult.

Adding a warning message when the file already exists or cannot be
created improves the visibility of file creation issues. This helps
developers identify and resolve file-related problems more easily.

Let's add a warning message in the verifyFileExistence method to notify
when the file already exists or cannot be created. This ensures that
developers are aware of potential issues during file creation.

The warning message is logged to the console, which is sufficient for
debugging purposes. This change does not affect the functionality of the
code but improves its maintainability and debugging experience.

meglwhy committed:
Add validation check for unmark command

The UnmarkCommand class did not previously validate whether the command
was explicitly 'unmark' before processing. This allowed potential
unexpected behavior in certain cases.

To improve clarity and prevent unintended execution, let's add a check
to verify if the command starts with 'unmark'. If it does, the program
now immediately throws an IOException with the message 'The method is
not done yet.' This ensures that the command execution is halted before
proceeding further.

This change enhances error handling and provides clear feedback to
developers when this method is called.

meglwhy committed:
Add validation for already marked tasks in MarkCommand

Previously, MarkCommand did not check whether a task was already marked
as done before attempting to mark it again. This could lead to redundant
operations and unnecessary writes to storage.

To improve error handling, let's add a check before marking a task as
done. If the task is already marked, the program now throws an IOException
with the message 'Task is already marked as done.' This prevents duplicate
operations and improves clarity for the user.

Additionally, this ensures that no unnecessary file writes occur if the
task was already marked.

meglwhy committed:
Improve README formatting and structure

The Koji Chatbot User Guide previously had a plain layout, making it harder
to read and navigate efficiently. Important commands lacked proper
highlighting, and sections were not clearly distinguished.

To enhance readability and user experience, let's:
- Add proper Markdown formatting, including bold headings and icons.
- Improve command highlighting using code blocks.
- Separate sections with clear dividers.
- Standardize expected output formatting for consistency.
- Make instructions more intuitive and engaging.

These changes make the README more structured, visually appealing, and
easier for users to follow when interacting with Koji Chatbot.

meglwhy committed:
Fix incorrect validation logic in UnmarkCommand

Previously, the UnmarkCommand constructor incorrectly checked if the
command was 'unmark' and immediately threw an IOException, making the
command unusable. This was not the intended behavior.

To fix this:
- Removed the incorrect validation that blocked execution for all 'unmark' commands.
- Ensured task indexing is properly validated before attempting to mark a task as not done.
- Improved error handling to provide accurate feedback to users.

With this fix, 'unmark' now correctly allows marking tasks as not done while
preventing invalid task numbers from being processed.
